pandas/core/arrays/datetimelike.py

```diff
@@ -478,14 +480,16 @@ def __setitem__(
             if lib.is_scalar(key):
                 raise ValueError("setting an array element with a sequence.")
 
-            if (not is_slice
-                    and len(key) != len(value)
-                    and not com.is_bool_indexer(key)):
-                msg = ("shape mismatch: value array of length '{}' does not "
-                       "match indexing result of length '{}'.")
-                raise ValueError(msg.format(len(key), len(value)))
-            if not is_slice and len(key) == 0:
-                return
+            if not is_slice:
+                key = cast(Sequence, key)
```

Comment:
Per discussions with @jreback we'd like to avoid the use of `cast` almost entirely. Would be preferable if you could just assign to a new variable within this branch

Reply:
> almost entirely
Are there instances in this PR where you'd like to keep `cast`? Otherwise, I'll work around it the handful of places is appears.

Reply:
If you can work around where it appears might be the most readable

Reply:
In this case, are you imagining a solution like this?
```py
if not is_slice:
    key_sequence = key
    if (len(key_sequence) != len(value)
            and not com.is_bool_indexer(key_sequence)):
        msg = ("shape mismatch: value array of length '{}' does "
                "not match indexing result of length '{}'.")
        raise ValueError(msg.format(
            len(key_sequence), len(value)))
    if len(key_sequence) == 0:
        return
```
That's how I read your comment above.

Just assigning a new variable leaves us with the same incompatible-type errors when calling `len()` that we started with. Using `key_sequence = key  # type: Sequence` gives us the error `Incompatible types in assignment (expression has type "Union[int, Sequence[int], Sequence[bool], slice]", variable has type "Sequence[Any]")`. The fundamental problem is that assigning to a new variable doesn't provide any new type information.

I haven't been able to think of a way to do this without at least one `cast` or a handful of ignores. As currently written, the single `cast` with the reassignment to `key` is about as unobtrusive as it gets. I think this is exactly what `cast` is for.

What's the reasoning for trying to avoid a construction like this?
```py
if not is_slice:
    key = cast(Sequence, key)
...
```
I'm sensitive to the idea that casting adds some cognitive overhead for authors and maintainers, but it's minimal in this case.

Reply:
Would doing something like `key_sequence = key # type: Sequence[bool]` work in the above? It sounds like the new error is the result of implicitly casting from a more refined `Sequence[bool]` (or int) in the union to a generic `Sequence`

Reply:
No, if the error message is put another away, the right side the equal sign is typed as `Union[int, Sequence[int], Sequence[bool], slice]` because that's what we told mypy `key` is in the `__setitem__` signature.

If we use something like `key_sequence = key # type: Sequence[bool]`, we're only telling mypy that `key_sequence` expects `Sequence[bool]`. We're not telling it anything about `key`, only `key_sequence`. That type comment doesn't narrow down `key`.

`cast(Sequence, key)`, on the other hand, explicitly tells mypy "key is a Sequence right here".

Reply:
The underlying error message, `Argument 1 to "len" has incompatible type "Union[int, Sequence[int], Sequence[bool], slice]"; expected "Sized"`, is really complaining because we're calling `len` on a variable that might be `int` or `slice`, neither of which support `len`. It's not really important for that error message what's inside the Sequence, only that the variable is definitely of a type that can respond to `len`.

Reply:
@gwrome is your last comment in reference to the static analysis or actual possible code paths?

Reply:
Static analysis. At that point in the actual execution, `slice` and `int` have already been handled. Mypy just isn't sophisticated enough to realize that.